### PR TITLE
MGMT-8277: add Kind completion metrics

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
Currently, we have a completion metric per YAML in the SRO CR
This PR adds a metric per Kind in the yamls of the SRO CR
So, that user will be able to see which OCP objects( DS, Roles, BuildConfig, etc')
his CR has created

### CODE REVIEW

File: pkg/resource/resource.go [hunk 1]
```diff
@@ -175,81 +176,111 @@ func CreateFromYAML(yamlFile []byte,
...
+	}
 
-		// We are only building a driver-container if we cannot pull the image
-		// We are asuming that vendors provide pre compiled DriverContainers
-		// If err == nil, build a new container, if err != nil skip it
-		if err = rebuildDriverContainer(obj); err != nil {
-			log.Info("Skipping building driver-container", "Name", obj.GetName())
-			return nil
-		}
+	log.Info("YEV - start parsing yaml for name, yaml", "name", name, "yaml", string(jsonSpec))
```
Comment (pacevedom) [inline pkg/resource/resource.go]: Forgot to remove this one?

File: pkg/resource/resource.go [hunk 2]
```diff
@@ -175,81 +176,111 @@ func CreateFromYAML(yamlFile []byte,
...
-		// kernel affinity related attributes only set if there is an
-		// annotation specialresource.openshift.io/kernel-affine: true
-		if kernel.IsObjectAffine(obj) {
-			if err = kernel.SetAffineAttributes(obj, kernelFullVersion, operatingSystemMajorMinor); err != nil {
-				return fmt.Errorf("cannot set kernel affine attributes: %w", err)
-			}
-		}
+	jsonSpec, err := yaml.YAMLToJSON(yamlSpec)
+	if err != nil {
+		return errors.Wrap(err, "Could not convert yaml file to json"+string(yamlSpec))
```
Comment (pacevedom) [inline pkg/resource/resource.go]: Prefer fmt.Errorf to wrap

File: pkg/resource/resource.go [hunk 3]
```diff
@@ -175,81 +176,111 @@ func CreateFromYAML(yamlFile []byte,
...
-			if strings.Contains(err.Error(), "failed calling webhook") {
-				return fmt.Errorf("webhook not ready, requeue: %w", err)
-			}
+	yamlKind := obj.GetKind()
+	metricValue := 0
+	defer func() {
+		metrics.SetCompletedKind(name, yamlKind, metricValue)
+	}()
+
+	//  Do not override the namespace if alreayd set
```
Comment (qbarrand) [inline pkg/resource/resource.go]:
```suggestion
	//  Do not override the namespace if already set
```

File: pkg/resource/resource.go (hunk 2 above)
Comment (qbarrand) [inline pkg/resource/resource.go]:
```suggestion
		return fmt.Errorf("could not convert yaml file to json: %s: %v", yamlSpec, err)
```

File: pkg/resource/resource.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: yes, thanks for noticing, removing

File: pkg/resource/resource.go [hunk 4]
```diff
@@ -175,81 +176,111 @@ func CreateFromYAML(yamlFile []byte,
...
+	}
+
+	// Callbacks before CRUD will update the manifests
+	if err = BeforeCRUD(obj, owner); err != nil {
+		return fmt.Errorf("before CRUD hooks failed: %w", err)
+	}
+	// Create Update Delete Patch resources
+	err = CRUD(obj, releaseInstalled, owner, name, namespace)
+	// The mutating webhook needs a couple of secs to be ready
+	// sleep for 5 secs and requeue
```
Comment (qbarrand) [inline pkg/resource/resource.go]: We don't actually sleep here.

File: pkg/resource/resource.go (hunk 2 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: fixed

File: pkg/resource/resource.go [hunk 5]
```diff
@@ -175,81 +176,109 @@ func CreateFromYAML(yamlFile []byte,
...
-		// We are asuming that vendors provide pre compiled DriverContainers
-		// If err == nil, build a new container, if err != nil skip it
-		if err = rebuildDriverContainer(obj); err != nil {
-			log.Info("Skipping building driver-container", "Name", obj.GetName())
-			return nil
-		}
+	yamlKind := obj.GetKind()
+	metricValue := 0
+	defer func() {
+		metrics.SetCompletedKind(name, yamlKind, metricValue)
```
Comment (qbarrand) [inline pkg/resource/resource.go]: If an SR creates two objects of the same kind and one is successfully created but not the other, wouldn't this result in the user getting unreliable information?

File: pkg/resource/resource.go [hunk 6]
```diff
@@ -175,81 +176,109 @@ func CreateFromYAML(yamlFile []byte,
...
+	}
+	// kernel affinity related attributes only set if there is an
+	// annotation specialresource.openshift.io/kernel-affine: true
+	if kernel.IsObjectAffine(obj) {
+		if err = kernel.SetAffineAttributes(obj, kernelFullVersion, operatingSystemMajorMinor); err != nil {
+			return fmt.Errorf("cannot set kernel affine attributes: %w", err)
 		}
+	}
+
+	// Add nodeSelector terms for the specialresource
```
Comment (qbarrand) [inline pkg/resource/resource.go]: Do I understand correctly that this selector is from the SR?
```suggestion
	// Add nodeSelector terms from the specialresource
```

File: pkg/resource/resource.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: good point, but i don't think that we will have two same kinds in the charts, very unlikely

File: pkg/resource/resource.go (hunk 6 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: the nodeSelector is defined in the CR spec, but the phrase can be under as add nodeSelector defined for the SR to the object, i will change the description accordingly

File: pkg/resource/resource.go (hunk 4 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: fixed

File: pkg/resource/resource.go (hunk 2 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: fixed

File: pkg/resource/resource.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: fixed

File: pkg/resource/resource.go (hunk 5 above)
Comment (qbarrand) [reply pkg/resource/resource.go]: `ping-pong` has two `Certificate` for example https://github.com/openshift/special-resource-operator/blob/master/charts/example/ping-pong-0.0.1/templates/0003_certificates.yaml

Would it make sense to also attach the resource name as a label? That would increase the metrics cardinality on Prometheus, but I don't think customers will be deploying many SpecialResources in their clusters anyway.

Comment (Veritasosrb) [reply pkg/resource/resource.go]: Chart can create multiple same kind of objects in different states or even in the same state yaml. Certificate, role, rolebinding, daemonset objects are created in multiple states in infoscale chart.

Comment (yevgeny-shnaidman) [reply pkg/resource/resource.go]: ok, thanks.
fixed

**Pull request merged**